### PR TITLE
[Enhancement] Sstable block cache and memory usage of LakePerisistentIndex

### PR DESCRIPTION
**Pull request by TszKitLo40**
## Why I'm doing:
1. The memory usage of LakePersistentIndex should be tracked. 
2. Sstable block cache should be supported in LakePersistentIndex which will be used in get api of sstable.
## What I'm doing:
1. Implement `memory_usage()` of LakePersistentIndex.
2. Support using block cache in sstable.

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/storage/lake/lake_persistent_index.cpp [hunk 1]
```diff
@@ -445,4 +446,12 @@ Status LakePersistentIndex::load_from_lake_tablet(TabletManager* tablet_mgr, con
     return Status::OK();
 }
 
+size_t LakePersistentIndex::memory_usage() const {
+    size_t mem_usage = _memtable->memory_usage();
```
Comment (luohaha) [inline be/src/storage/lake/lake_persistent_index.cpp:450]: Will `_memtable` be nullptr?

File: be/src/storage/lake/update_manager.cpp [hunk 2]
```diff
@@ -622,6 +636,7 @@ Status UpdateManager::publish_primary_compaction(const TxnLogPB_OpCompaction& op
         delvecs.emplace_back(rssid, dv);
         compaction_state.release_segments(i);
     }
+    _block_cache->update_memory_usage();
 
     // 3. update TabletMeta and write to meta file
     for (auto&& each : delvecs) {
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/lake/update_manager.cpp:642]: follow this one.

File: be/src/storage/lake/update_manager.cpp [hunk 3]
```diff
@@ -52,6 +52,8 @@ UpdateManager::UpdateManager(LocationProvider* location_provider, MemTracker* me
     _update_state_cache.set_mem_tracker(_update_state_mem_tracker.get());
 
     _index_cache.set_capacity(update_mem_limit);
+    _block_cache = std::make_unique<PersistentIndexBlockCache>(
+            mem_tracker, update_mem_limit * config::lake_pk_index_block_cache_limit);
```
Comment (luohaha) [inline be/src/storage/lake/update_manager.cpp:56]: Unified the config usage here. Maybe `lake_pk_index_block_cache_limit_percent = 10` is better.

File: be/src/storage/lake/lake_persistent_index.cpp (hunk 1 above)
Comment (TszKitLo40) [reply be/src/storage/lake/lake_persistent_index.cpp:462]: It will be constructed in constructor of LakePersistentIndex.

Comment (luohaha) [reply be/src/storage/lake/lake_persistent_index.cpp:462]: It's best not to make optimistic assumptions

File: be/src/storage/lake/update_manager.cpp (hunk 3 above)
Comment (luohaha) [reply be/src/storage/lake/update_manager.cpp:56]: And create a new tracker under `_update_mem_tracker`, and set name to it.

File: be/src/storage/lake/update_manager.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (TszKitLo40) [reply be/src/storage/lake/update_manager.cpp:650]: done

File: be/src/storage/lake/update_manager.cpp (hunk 3 above)
Comment (TszKitLo40) [reply be/src/storage/lake/update_manager.cpp]: done

File: be/src/storage/lake/update_manager.h [hunk 4]
```diff
@@ -206,8 +223,11 @@ class UpdateManager {
     std::unique_ptr<MemTracker> _index_cache_mem_tracker;
     std::unique_ptr<MemTracker> _update_state_mem_tracker;
     std::unique_ptr<MemTracker> _compaction_state_mem_tracker;
+    std::unique_ptr<MemTracker> _block_cache_mem_tracker;
```
Comment (luohaha) [inline be/src/storage/lake/update_manager.h:226]: Move `_block_cache_mem_tracker` to `_block_cache`.

File: be/src/storage/lake/update_manager.cpp [hunk 5]
```diff
@@ -52,6 +52,12 @@ UpdateManager::UpdateManager(LocationProvider* location_provider, MemTracker* me
     _update_state_cache.set_mem_tracker(_update_state_mem_tracker.get());
 
     _index_cache.set_capacity(update_mem_limit);
+
+    const int64_t block_cache_mem_limit =
+            update_mem_limit * std::max(std::min(100, config::lake_pk_index_block_cache_limit_percent), 0) / 100;
+    _block_cache_mem_tracker =
+            std::make_unique<MemTracker>(block_cache_mem_limit, "lake_persistent_index_block_cache", mem_tracker);
+    _block_cache = std::make_unique<PersistentIndexBlockCache>(_block_cache_mem_tracker.get(), block_cache_mem_limit);
```
Comment (luohaha) [inline be/src/storage/lake/update_manager.cpp:60]: _block_cache = std::make_unique<PersistentIndexBlockCache>(std::make_unique<MemTracker>(block_cache_mem_limit, "lake_persistent_index_block_cache", mem_tracker));

Comment (luohaha) [reply be/src/storage/lake/update_manager.cpp:60]: Move `_block_cache_mem_tracker` to `_block_cache`.

File: be/src/storage/lake/update_manager.h [hunk 6]
```diff
@@ -52,6 +53,20 @@ class LakeDelvecLoader : public DelvecLoader {
...
+    explicit PersistentIndexBlockCache(MemTracker* mem_tracker, int64_t cache_limit);
+
+    void update_memory_usage();
+
+    Cache* cache() { return _cache.get(); }
+
+private:
+    std::atomic<size_t> _memory_usage{0};
+    std::unique_ptr<Cache> _cache;
+    std::unique_ptr<MemTracker> _mem_tracker;
```
Comment (luohaha) [inline be/src/storage/lake/update_manager.h:67]: Remove this.

**Review by luohaha — DISMISSED** [review]

File: be/src/common/config.h [hunk 7]
```diff
@@ -982,6 +982,7 @@ CONF_mInt64(lake_pk_compaction_min_input_segments, "5");
 CONF_mInt32(lake_pk_preload_memory_limit_percent, "30");
 CONF_mInt32(lake_pk_index_sst_min_compaction_versions, "2");
 CONF_mInt32(lake_pk_index_sst_max_compaction_versions, "5");
+CONF_mInt32(lake_pk_index_block_cache_limit_percent, "10");
```
Comment (sduzh) [inline be/src/common/config.h:985]: This config was defined as mutable but modify it at runtime does not take any effect right?

**Review by sduzh — APPROVED** [review]

**Review by luohaha — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Pull request merged**